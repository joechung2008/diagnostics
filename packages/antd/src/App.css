html,
body {
  height: 100%;
  margin: 0;
  padding: 0;
}

*,
*::after,
*::before {
  box-sizing: border-box;
}

.centered {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  width: 100%;
}

.extension-root {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
  max-height: max-content;
  overflow-y: auto;
}

.flexbox {
  display: flex;
  flex-direction: column;
  height: 100vh;
}

.grow {
  flex: 1;
}

.stack {
  display: flex;
  flex-direction: row;
  height: 100%;
}

.tab-list {
  padding-left: 1rem;
}

.tab-panel {
  flex: 1;
  overflow-y: auto;
}
