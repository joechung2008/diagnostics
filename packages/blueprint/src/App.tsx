import {
  Button,
  ButtonGroup,
  NonIdealState,
  Spinner,
  Tab,
  Tabs,
} from "@blueprintjs/core";
import clsx from "clsx";
import { useCallback, useMemo, useState } from "react";
import useSWR from "swr";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { isExtensionInfo } from "./utils";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl =
  | "https://hosting.portal.azure.net/api/diagnostics"
  | "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics"
  | "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics";

interface AppProps {
  className?: string;
}

const App = ({ className }: AppProps) => {
  const [environment, setEnvironment] = useState<EnvironmentUrl>(
    Environment.Public
  );
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR(environment, async (environment: string) => {
    const response = await fetch(environment);
    if (!response.ok) {
      throw new Error(`Failed to fetch diagnostics: ${response.statusText}`);
    }

    return response.json();
  });

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const handleLinkClick = useCallback(
    (item?: KeyedNavLink) => {
      if (item) {
        const extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo(extension)) {
          setExtension(extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  return (
    <div className={clsx("flexbox", className)}>
      <ButtonGroup>
        <Button
          active={environment === Environment.Public}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Public);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Public Cloud
        </Button>
        <Button
          active={environment === Environment.Fairfax}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Fairfax);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Fairfax
        </Button>
        <Button
          active={environment === Environment.Mooncake}
          variant="minimal"
          onClick={() => {
            setEnvironment(Environment.Mooncake);
            setExtension(undefined);
            setSelectedTab("extensions");
          }}
        >
          Mooncake
        </Button>
        <span className="button-group-separator" />
        {showPaasServerless && (
          <Button
            variant="minimal"
            onClick={() => {
              const paasserverless = diagnostics?.extensions["paasserverless"];
              if (isExtensionInfo(paasserverless)) {
                setExtension(paasserverless);
                setSelectedTab("extensions");
              }
            }}
          >
            paasserverless
          </Button>
        )}
        <Button
          variant="minimal"
          onClick={() => {
            const websites = diagnostics?.extensions["websites"];
            if (isExtensionInfo(websites)) {
              setExtension(websites);
              setSelectedTab("extensions");
            }
          }}
        >
          websites
        </Button>
      </ButtonGroup>
      <Tabs
        selectedTabId={selectedTab}
        size="large"
        onChange={(tabId) => setSelectedTab(tabId as string)}
      >
        <Tab
          id="extensions"
          aria-controls="extensions-tab"
          title="Extensions"
        />
        <Tab id="build" aria-controls="build-tab" title="Build Information" />
        <Tab
          id="server"
          aria-controls="server-tab"
          title="Server Information"
        />
      </Tabs>
      {isLoading ? (
        <div className="tab-panel vertical-center">
          <Spinner aria-label="Loading..." />
        </div>
      ) : error ? (
        <NonIdealState icon="error" title="Error" description={error.message} />
      ) : selectedTab === "extensions" && diagnostics?.extensions ? (
        <div id="extensions-tab" className="tab-panel">
          <div className="stack">
            <Extensions
              extensions={diagnostics.extensions}
              onLinkClick={handleLinkClick}
            />
            {extension && <Extension {...extension} />}
          </div>
        </div>
      ) : selectedTab === "build" && diagnostics?.buildInfo ? (
        <div id="build-tab" className="tab-panel">
          <BuildInfo {...diagnostics.buildInfo} />
        </div>
      ) : selectedTab === "server" && diagnostics?.serverInfo ? (
        <div id="server-tab" className="tab-panel">
          <ServerInfo {...diagnostics.serverInfo} />
        </div>
      ) : null}
    </div>
  );
};

export default App;
