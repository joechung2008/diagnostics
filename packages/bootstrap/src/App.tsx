import { Activity, useCallback, useMemo, useState } from "react";
import useSWR from "swr";
import {
  Alert,
  Button,
  ButtonGroup,
  Dropdown,
  Nav,
  Spinner,
} from "react-bootstrap";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { isExtensionInfo } from "./utils";
import { useSystemTheme } from "./useSystemTheme";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl = (typeof Environment)[keyof typeof Environment];

const App: React.FC = () => {
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [environment, setEnvironment] = useState<EnvironmentUrl>(
    Environment.Public
  );
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  // Use system theme hook to set data-bs-theme on html element
  useSystemTheme();

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR<Diagnostics>(environment, async (environment: string) => {
    const response = await fetch(environment);
    if (!response.ok) {
      throw new Error(`Error fetching diagnostics: ${response.statusText}`);
    }

    return response.json();
  });

  const environmentName = useMemo(() => {
    switch (environment) {
      case Environment.Public:
        return "Public Cloud";
      case Environment.Fairfax:
        return "Fairfax";
      case Environment.Mooncake:
        return "Mooncake";
      default:
        return "Select environment";
    }
  }, [environment]);

  const environments = useMemo(
    () => [
      {
        key: "public",
        text: "Public Cloud",
        selected: environment === Environment.Public,
        onClick: () => {
          setEnvironment(Environment.Public);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "fairfax",
        text: "Fairfax",
        selected: environment === Environment.Fairfax,
        onClick: () => {
          setEnvironment(Environment.Fairfax);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "mooncake",
        text: "Mooncake",
        selected: environment === Environment.Mooncake,
        onClick: () => {
          setEnvironment(Environment.Mooncake);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
    ],
    [environment]
  );

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions?.["paasserverless"]),
    [diagnostics]
  );

  const handleLinkClick = useCallback(
    (_?: React.MouseEvent, item?: KeyedNavLink) => {
      if (item) {
        const extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo(extension)) {
          setExtension(extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  return (
    <div className="flexbox">
      <ButtonGroup className="w-auto align-self-start">
        <Dropdown>
          <Dropdown.Toggle variant="outline-primary" id="environment-dropdown">
            {environmentName}
          </Dropdown.Toggle>
          <Dropdown.Menu>
            {environments.map((env) => (
              <Dropdown.Item
                key={env.key}
                active={env.selected}
                onClick={env.onClick}
              >
                {env.text}
              </Dropdown.Item>
            ))}
          </Dropdown.Menu>
        </Dropdown>
        {showPaasServerless && (
          <Button
            variant="outline-secondary"
            onClick={() => {
              const paasserverless = diagnostics?.extensions["paasserverless"];
              if (isExtensionInfo(paasserverless)) {
                setExtension(paasserverless);
                setSelectedTab("extensions");
              }
            }}
          >
            paasserverless
          </Button>
        )}
        <Button
          variant="outline-secondary"
          onClick={() => {
            const websites = diagnostics?.extensions["websites"];
            if (isExtensionInfo(websites)) {
              setExtension(websites);
              setSelectedTab("extensions");
            }
          }}
        >
          websites
        </Button>
      </ButtonGroup>
      <Nav
        variant="tabs"
        activeKey={selectedTab}
        onSelect={(selectedKey) => setSelectedTab(selectedKey || "extensions")}
      >
        <Nav.Item>
          <Nav.Link eventKey="extensions">Extensions</Nav.Link>
        </Nav.Item>
        <Nav.Item>
          <Nav.Link eventKey="build">Build Information</Nav.Link>
        </Nav.Item>
        <Nav.Item>
          <Nav.Link eventKey="server">Server Information</Nav.Link>
        </Nav.Item>
      </Nav>
      <Activity mode={isLoading && !error ? "visible" : "hidden"}>
        <div className="tab-panel centered">
          <Spinner animation="border" />
        </div>
      </Activity>
      <Activity mode={error ? "visible" : "hidden"}>
        <div className="tab-panel centered">
          <Alert variant="danger">
            An error occurred while loading diagnostics.
          </Alert>
        </div>
      </Activity>
      <Activity
        mode={
          selectedTab === "extensions" && !error && !isLoading
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.extensions && (
          <div className="tab-panel">
            <div className="stack">
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </div>
          </div>
        )}
      </Activity>
      <Activity
        mode={
          selectedTab === "build" && !error && !isLoading ? "visible" : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <div className="tab-panel">
            <BuildInfo {...diagnostics.buildInfo} />
          </div>
        )}
      </Activity>
      <Activity
        mode={
          selectedTab === "server" && !error && !isLoading
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.serverInfo && (
          <div className="tab-panel">
            <ServerInfo {...diagnostics!.serverInfo} />
          </div>
        )}
      </Activity>
    </div>
  );
};

export default App;
