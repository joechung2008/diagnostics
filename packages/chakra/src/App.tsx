import BuildInfo from "@/BuildInfo";
import { ColorModeButton } from "@/components/ui/ColorMode";
import Extension from "@/Extension";
import Extensions from "@/Extensions";
import ServerInfo from "@/ServerInfo";
import { isExtensionInfo } from "@/utils";
import {
  Box,
  Button,
  Flex,
  Menu,
  Portal,
  Tabs,
  Alert,
  Spinner,
} from "@chakra-ui/react";
import {
  startTransition,
  useCallback,
  useMemo,
  useState,
  Activity,
} from "react";
import useSWR from "swr";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl = (typeof Environment)[keyof typeof Environment];

function getEnvironnmentName(environment: EnvironmentUrl | undefined): string {
  switch (environment) {
    case Environment.Public:
      return "Public Cloud";
    case Environment.Fairfax:
      return "Fairfax";
    case Environment.Mooncake:
      return "Mooncake";
    default:
      return "Select environment";
  }
}

const App: React.FC = () => {
  const [environments, setEnvironments] = useState<
    Record<string, EnvironmentUrl[]>
  >({
    environment: [Environment.Public],
  });
  const environment = useMemo<EnvironmentUrl>(
    () => environments.environment[0],
    [environments.environment]
  );
  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR(environment, async (url: string): Promise<Diagnostics> => {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch diagnostics: ${response.statusText}`);
    }

    return response.json();
  });
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const handleLinkClick: ExtensionsProps["onLinkClick"] = useCallback(
    (_, item) => {
      if (item) {
        const $extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo($extension)) {
          setExtension($extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environmentName = useMemo(
    () => getEnvironnmentName(environment),
    [environment]
  );

  const handleEnvironmentChange = useCallback((value: string) => {
    if (value) {
      startTransition(() => {
        setEnvironments((previous) => ({
          ...previous,
          environment: [value as EnvironmentUrl],
        }));
        setExtension(undefined);
        setSelectedTab("extensions");
      });
    }
  }, []);

  return (
    <Flex flexDirection="column" h="100vh" p="1">
      <Flex justify="space-between">
        <Flex gap="1">
          <Menu.Root>
            <Menu.Trigger asChild>
              <Button variant="outline">{environmentName}</Button>
            </Menu.Trigger>
            <Portal>
              <Menu.Positioner>
                <Menu.Content>
                  {Object.entries(Environment).map(([key, value]) => (
                    <Menu.Item
                      key={key}
                      value={value}
                      onClick={() => handleEnvironmentChange(value)}
                    >
                      {getEnvironnmentName(value)}
                    </Menu.Item>
                  ))}
                </Menu.Content>
              </Menu.Positioner>
            </Portal>
          </Menu.Root>
          <Activity mode={showPaasServerless ? "visible" : "hidden"}>
            <Button
              key="paasserverless"
              variant="ghost"
              onClick={() => {
                const paasserverless =
                  diagnostics?.extensions["paasserverless"];
                if (isExtensionInfo(paasserverless)) {
                  setExtension(paasserverless);
                  setSelectedTab("extensions");
                }
              }}
            >
              paasserverless
            </Button>
          </Activity>
          <Button
            key="websites"
            variant="ghost"
            onClick={() => {
              const websites = diagnostics?.extensions["websites"];
              if (isExtensionInfo(websites)) {
                setExtension(websites);
                setSelectedTab("extensions");
              }
            }}
          >
            websites
          </Button>
        </Flex>
        <ColorModeButton />
      </Flex>
      <Tabs.Root
        value={selectedTab}
        onValueChange={(details) => setSelectedTab(details.value)}
      >
        <Tabs.List>
          <Tabs.Trigger aria-controls="extensions-tab" value="extensions">
            Extensions
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="build-tab" value="build">
            Build Information
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="server-tab" value="server">
            Server Information
          </Tabs.Trigger>
        </Tabs.List>
      </Tabs.Root>
      <Activity mode={isLoading && !error ? "visible" : "hidden"}>
        <Box
          alignItems="center"
          display="flex"
          flex="1"
          justifyContent="center"
        >
          <Spinner aria-label="Loading..." />
        </Box>
      </Activity>
      <Activity mode={!isLoading && error ? "visible" : "hidden"}>
        <Box
          alignItems="center"
          display="flex"
          flex="1"
          justifyContent="center"
        >
          <Alert.Root status="error" maxW="md">
            <Alert.Indicator />
            <Alert.Title>Error</Alert.Title>
            <Alert.Description>{error?.message}</Alert.Description>
          </Alert.Root>
        </Box>
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "extensions"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.extensions && (
          <Box id="extensions-tab" flex="1" overflowY="auto" role="tabpanel">
            <Flex flexDirection="row" gap="4" h="100%">
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </Flex>
          </Box>
        )}
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "build" ? "visible" : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <Box id="build-tab" flex="1" overflowY="auto" role="tabpanel">
            <BuildInfo {...diagnostics.buildInfo} />
          </Box>
        )}
      </Activity>
      <Activity
        mode={
          !isLoading && !error && selectedTab === "server"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.serverInfo && (
          <Box id="server-tab" flex="1" overflowY="auto" role="tabpanel">
            <ServerInfo {...diagnostics.serverInfo} />
          </Box>
        )}
      </Activity>
    </Flex>
  );
};

export default App;
