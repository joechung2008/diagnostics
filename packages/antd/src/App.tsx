import { Empty, Flex, Menu, Spin, Tabs } from "antd";
import { Activity, useCallback, useMemo, useState } from "react";
import useSWR from "swr";
import BuildInfo from "./BuildInfo";
import Extension from "./Extension";
import Extensions from "./Extensions";
import ServerInfo from "./ServerInfo";
import { isExtensionInfo, when } from "./utils";

import "./App.css";

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

type EnvironmentUrl = (typeof Environment)[keyof typeof Environment];

const items = [
  { key: "extensions", label: "Extensions" },
  { key: "build", label: "Build Information" },
  { key: "server", label: "Server Information" },
];

const App: React.FC = () => {
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [environment, setEnvironment] = useState<EnvironmentUrl>(
    Environment.Public
  );
  const [selectedTab, setSelectedTab] = useState("extensions");

  const {
    data: diagnostics,
    error,
    isLoading,
  } = useSWR<Diagnostics>(environment, async (environment: string) => {
    const response = await fetch(environment);

    if (!response.ok) {
      throw new Error(`Error fetching diagnostics: ${response.statusText}`);
    }

    return response.json();
  });

  const environmentName = useMemo(() => {
    switch (environment) {
      case Environment.Public:
        return "Public Cloud";
      case Environment.Fairfax:
        return "Fairfax";
      case Environment.Mooncake:
        return "Mooncake";
      default:
        return "Select environment";
    }
  }, [environment]);

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environments = useMemo(
    () => [
      {
        key: "public",
        text: "Public Cloud",
        selected: environment === Environment.Public,
        onClick: () => {
          setEnvironment(Environment.Public);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "fairfax",
        text: "Fairfax",
        selected: environment === Environment.Fairfax,
        onClick: () => {
          setEnvironment(Environment.Fairfax);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
      {
        key: "mooncake",
        text: "Mooncake",
        selected: environment === Environment.Mooncake,
        onClick: () => {
          setEnvironment(Environment.Mooncake);
          setExtension(undefined);
          setSelectedTab("extensions");
        },
      },
    ],
    [environment]
  );

  const menuItems = useMemo(
    () => [
      {
        key: "env-dropdown",
        label: environmentName,
        children: environments.map((env) => ({
          key: env.key,
          label: env.text,
          onClick: env.onClick,
        })),
      },
      ...when(showPaasServerless, {
        key: "paasserverless",
        label: "paasserverless",
        onClick: () => {
          const paasserverless = diagnostics?.extensions["paasserverless"];
          if (isExtensionInfo(paasserverless)) {
            setExtension(paasserverless);
            setSelectedTab("extensions");
          }
        },
      }),
      {
        key: "websites",
        label: "websites",
        onClick: () => {
          const websites = diagnostics?.extensions["websites"];
          if (isExtensionInfo(websites)) {
            setExtension(websites);
            setSelectedTab("extensions");
          }
        },
      },
    ],
    [environments, environmentName, showPaasServerless, diagnostics?.extensions]
  );

  const handleLinkClick = useCallback(
    (_?: React.MouseEvent, item?: KeyedNavLink) => {
      if (item) {
        const extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo(extension)) {
          setExtension(extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  return (
    <div className="flexbox">
      <Menu mode="horizontal" items={menuItems} />
      <Tabs
        activeKey={selectedTab}
        className="tab-list"
        items={items}
        onChange={(key) => setSelectedTab(key)}
      />
      <Activity mode={isLoading ? "visible" : "hidden"}>
        <div className="tab-panel">
          <Flex className="centered">
            <Spin size="large" />
          </Flex>
        </div>
      </Activity>
      <Activity mode={error && !isLoading ? "visible" : "hidden"}>
        {error && (
          <div className="tab-panel">
            <Flex className="centered">
              <Empty
                description={`Error: ${error.message ?? "Unexpected error"}`}
              />
            </Flex>
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "extensions"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.extensions && (
          <div className="tab-panel">
            <Flex className="stack">
              <Extensions
                extensions={diagnostics.extensions}
                onLinkClick={handleLinkClick}
              />
              {extension && <Extension {...extension} />}
            </Flex>
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "build" ? "visible" : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <div className="tab-panel">
            <BuildInfo {...diagnostics.buildInfo} />
          </div>
        )}
      </Activity>
      <Activity
        mode={
          !error && !isLoading && selectedTab === "server"
            ? "visible"
            : "hidden"
        }
      >
        {diagnostics?.buildInfo && (
          <div className="tab-panel">
            <ServerInfo {...diagnostics.serverInfo} />
          </div>
        )}
      </Activity>
    </div>
  );
};

export default App;
