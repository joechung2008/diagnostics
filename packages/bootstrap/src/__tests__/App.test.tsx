import { render } from "@testing-library/react";
import useSWR from "swr";
import { describe, expect, it, vi } from "vitest";
import App from "../App";

vi.mock("swr");

// Mock matchMedia for theme detection
const mockMatchMedia = vi.fn().mockImplementation((query: string) => ({
  matches: false, // Default to light theme
  media: query,
  onchange: null,
  addListener: vi.fn(),
  removeListener: vi.fn(),
  addEventListener: vi.fn(),
  removeEventListener: vi.fn(),
  dispatchEvent: vi.fn(),
}));

vi.stubGlobal("matchMedia", mockMatchMedia);

describe("App", () => {
  const mockDiagnostics = {
    buildInfo: { buildVersion: "1.0.0" },
    extensions: {
      websites: { extensionName: "websites" },
      paasserverless: { extensionName: "paasserverless" },
      testExt: { extensionName: "testExt" },
    },
    serverInfo: {
      hostname: "test-server",
      deploymentId: "123",
      extensionSync: { totalSyncAllCount: 5 },
      nodeVersions: "18.0.0",
      serverId: "server-1",
      uptime: 3600,
    },
  };

  it("shows loading spinner while loading diagnostics", () => {
    vi.mocked(useSWR).mockReturnValue({
      data: undefined,
      error: undefined,
      isLoading: true,
      isValidating: false,
      mutate: vi.fn(),
    });

    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows error alert when diagnostics could not be loaded", () => {
    vi.mocked(useSWR).mockReturnValue({
      data: undefined,
      error: new Error("Failed to load"),
      isLoading: false,
      isValidating: false,
      mutate: vi.fn(),
    });

    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows UI when diagnostics are loaded", () => {
    vi.mocked(useSWR).mockReturnValue({
      data: mockDiagnostics,
      error: undefined,
      isLoading: false,
      isValidating: false,
      mutate: vi.fn(),
    });

    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
