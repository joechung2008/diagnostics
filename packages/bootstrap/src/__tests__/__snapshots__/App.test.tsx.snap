// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > shows UI when diagnostics are loaded 1`] = `
<div
  class="flexbox"
>
  <div
    class="w-auto align-self-start btn-group"
    role="group"
  >
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle btn btn-outline-primary"
        id="environment-dropdown"
        type="button"
      >
        Public Cloud
      </button>
    </div>
    <button
      class="btn btn-outline-secondary"
      type="button"
    >
      paasserverless
    </button>
    <button
      class="btn btn-outline-secondary"
      type="button"
    >
      websites
    </button>
  </div>
  <div
    class="nav nav-tabs"
  >
    <div
      class="nav-item"
    >
      <a
        class="nav-link active"
        data-rr-ui-event-key="extensions"
        href="#"
        role="button"
        tabindex="0"
      >
        Extensions
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="build"
        href="#"
        role="button"
        tabindex="0"
      >
        Build Information
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="server"
        href="#"
        role="button"
        tabindex="0"
      >
        Server Information
      </a>
    </div>
  </div>
  <div
    class="tab-panel centered"
    style="display: none;"
  >
    <div
      class="spinner-border"
    />
  </div>
  <div
    class="tab-panel centered"
    style="display: none;"
  >
    <div
      class="fade alert alert-danger show"
      role="alert"
    >
      An error occurred while loading diagnostics.
    </div>
  </div>
  <div
    class="tab-panel"
  >
    <div
      class="stack"
    >
      <div
        class="extension-root overflow-x-hidden"
      >
        <div
          aria-label="Extensions"
          class="flex-column nav nav-pills"
        >
          <a
            class="nav-link"
            href="#"
            role="button"
            tabindex="0"
          >
            paasserverless
          </a>
          <a
            class="nav-link"
            href="#"
            role="button"
            tabindex="0"
          >
            testExt
          </a>
          <a
            class="nav-link"
            href="#"
            role="button"
            tabindex="0"
          >
            websites
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    class="tab-panel"
    style="display: none;"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Build Version
          </td>
          <td>
            1.0.0
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="tab-panel"
    style="display: none;"
  >
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Hostname
          </td>
          <td>
            test-server
          </td>
        </tr>
        <tr>
          <td>
            Uptime
          </td>
          <td>
            3600
          </td>
        </tr>
        <tr>
          <td>
            Server ID
          </td>
          <td>
            server-1
          </td>
        </tr>
        <tr>
          <td>
            Deployment ID
          </td>
          <td>
            123
          </td>
        </tr>
        <tr>
          <td>
            Node Versions
          </td>
          <td>
            18.0.0
          </td>
        </tr>
        <tr>
          <td>
            Extension Sync | Total Sync All Count
          </td>
          <td>
            5
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`App > shows error alert when diagnostics could not be loaded 1`] = `
<div
  class="flexbox"
>
  <div
    class="w-auto align-self-start btn-group"
    role="group"
  >
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle btn btn-outline-primary"
        id="environment-dropdown"
        type="button"
      >
        Public Cloud
      </button>
    </div>
    <button
      class="btn btn-outline-secondary"
      type="button"
    >
      websites
    </button>
  </div>
  <div
    class="nav nav-tabs"
  >
    <div
      class="nav-item"
    >
      <a
        class="nav-link active"
        data-rr-ui-event-key="extensions"
        href="#"
        role="button"
        tabindex="0"
      >
        Extensions
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="build"
        href="#"
        role="button"
        tabindex="0"
      >
        Build Information
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="server"
        href="#"
        role="button"
        tabindex="0"
      >
        Server Information
      </a>
    </div>
  </div>
  <div
    class="tab-panel centered"
    style="display: none;"
  >
    <div
      class="spinner-border"
    />
  </div>
  <div
    class="tab-panel centered"
  >
    <div
      class="fade alert alert-danger show"
      role="alert"
    >
      An error occurred while loading diagnostics.
    </div>
  </div>
</div>
`;

exports[`App > shows loading spinner while loading diagnostics 1`] = `
<div
  class="flexbox"
>
  <div
    class="w-auto align-self-start btn-group"
    role="group"
  >
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle btn btn-outline-primary"
        id="environment-dropdown"
        type="button"
      >
        Public Cloud
      </button>
    </div>
    <button
      class="btn btn-outline-secondary"
      type="button"
    >
      websites
    </button>
  </div>
  <div
    class="nav nav-tabs"
  >
    <div
      class="nav-item"
    >
      <a
        class="nav-link active"
        data-rr-ui-event-key="extensions"
        href="#"
        role="button"
        tabindex="0"
      >
        Extensions
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="build"
        href="#"
        role="button"
        tabindex="0"
      >
        Build Information
      </a>
    </div>
    <div
      class="nav-item"
    >
      <a
        class="nav-link"
        data-rr-ui-event-key="server"
        href="#"
        role="button"
        tabindex="0"
      >
        Server Information
      </a>
    </div>
  </div>
  <div
    class="tab-panel centered"
  >
    <div
      class="spinner-border"
    />
  </div>
  <div
    class="tab-panel centered"
    style="display: none;"
  >
    <div
      class="fade alert alert-danger show"
      role="alert"
    >
      An error occurred while loading diagnostics.
    </div>
  </div>
</div>
`;
